src/main/java/ch/njol/skript/effects/EffWakeupSleep.java

```diff
@@ -77,26 +76,38 @@ public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelaye
 	@Override
 	protected void execute(Event event) {
 		Location location = null;
-		if (this.location != null) {
+		if (this.location != null)
 			location = this.location.getSingle(event);
-			if (location == null)
-				warning("The provided location is not set. This effect will have no effect for villagers and players.");
-		}
+		boolean warned = false;
 		for (LivingEntity entity : entities.getArray(event)) {
 			if (entity instanceof Bat bat) {
 				bat.setAwake(!sleep);
 			} else if (entity instanceof Villager villager) {
+				if (sleep && location == null) {
+					if (!warned) {
+						warned = true;
+						warning("The provided location is not set. This effect will have no effect for villagers and players.");
+					}
+					continue;
```

Comment:
might be nice to de-duplicate this by just setting warning to true and checking the state of warning after the loop.